tools/verify_images.py

```diff
@@ -0,0 +1,118 @@
+#!/usr/bin/env python3
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2024
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Utility script to verify that all images have alt text"""
+
+from pathlib import Path
+import multiprocessing
+import sys
+import glob
+
+# Dictionary to allowlist lines of code that the checker will not error
+# Format: {"file_path": [list_of_line_numbers]}
+ALLOWLIST_MISSING_ALT_TEXT = {
+    "qiskit_ibm_runtime/fake_provider/__init__.py": [34, 55, 63],
+    "qiskit_ibm_runtime/transpiler/passes/scheduling/dynamical_decoupling.py": [56, 88],
+    "qiskit_ibm_runtime/transpiler/passes/scheduling/__init__.py": [
+        51,
+        102,
+        137,
+        152,
+        176,
+        207,
+        226,
+        240,
+        256,
+        274,
+        294,
+        313,
+        328,
+        342,
+        363,
+        382,
+        406,
+    ],
+}
+
+
+def is_image(line: str) -> bool:
+    return line.strip().startswith((".. image:", ".. plot:"))
+
+
+def in_allowlist(filename: str, line_num: int) -> bool:
+    return line_num in ALLOWLIST_MISSING_ALT_TEXT.get(filename, [])
+
+
+def validate_image(file_path: str) -> tuple[str, list[str]]:
+    """Validate all the images of a single file"""
+    invalid_images: list[str] = []
+
+    lines = Path(file_path).read_text().splitlines()
+
+    for line_number, line in enumerate(lines):
+        if not is_image(line) or in_allowlist(file_path, line_number + 1):
+            continue
+
+        options: list[str] = []
+        options_line_number = line_number
```

Comment:
It might be possible to get rid of this and simply use `line_index`? Right now, you end up iterating over the same line twice because `line_index` is no being incremented after identifying that these option lines are not images. You'd replace the `for` loop with a `while` loop sith the same condition as your inner loop on 69.

Imagine this:

```
[line 1] .. image::
[line 2]
[line 3] .. image::
```

The options code would increment `line_index` to 2 and then break out of the inner loop because there were no options.

I think the only edge case to be careful about is when you have a new image as the next line. Imagine this:

```
[line 1] .. image::
[line 2] .. image::
```

Your options code stops at line 2 because there were no options for the first image. You can break out of the inner loop now. Things will work properly as long as the outer loop doesn't increment `line_index` when entering the next iteration: you need to start the iteration at line 2 and not try going to line 3.

Reply:
That's an interesting suggestion, thanks! I think we can't remove the `options_line_number` because we need a temporary variable to store the line of the image for the error message. However, we are indeed iterating twice in some lines. I chose that approach because I thought it was more readable and had in mind that the number of options of a directive (:= lines iterated twice) is very small, at most 4/5 I would say.

Regardless, I think it's a good idea to try to iterate only once. If I understand correctly what you are suggesting, I think it would be something like this:

```python
    line_index = 0
    while line_index + 1 < len(lines):
        line = lines[line_index].strip()
        if not is_image(line) or in_allowlist(file_path, line_index + 1):
            line_index += 1
            continue

        options: list[str] = []
        line_index_tmp = line_index

        while line_index + 1 < len(lines):
            line_index += 1
            option_line = lines[line_index].strip()

            if not option_line.startswith(":"):
                break

            options.append(option_line)

         `Check if the image is valid or not...`
```

I think we could iterate once and at the same time remove the inner loop by doing something like this, defining some helper functions:

```python
    line_index = 0
    image_found = False
    image_line = -1
    options: list[str] = []

    while line_index < len(lines):
        line = lines[line_index].strip()

        if image_found and not is_option(line) and not is_valid_image(options):
            invalid_images.append(f"- Error in line {image_line}: {lines[image_line-1].strip()}")
            image_found = False
            options = []
            continue

        if image_found and is_option(line):
            options.append(line)
            line_index += 1
            continue

        image_found = is_image(line) and not in_allowlist(file_path, line_index + 1)
        image_line = line_index + 1
        line_index += 1
```

Do you think either of these approaches is preferable over the current one taking into account performance/readability?

Reply:
Thank you for the suggestion! I have been looking at the code, and I think we could even remove the inner loop and still iterate only once per line. For each line, we only need two special cases. If we are looking at an image option we just need to save it, and if there are no more options after seeing an image we create an error message.

```python
    image_found = False
    options: list[str] = []

    for line_index, line in enumerate(lines):
        if image_found and is_option(line):
            options.append(line)
            continue

        if image_found and not is_valid_image(options):
            image_line = line_index - len(options)
            invalid_images.append(f"- Error in line {image_line}: {lines[image_line-1].strip()}")

        image_found = is_image(line)
        options = []
```